### PR TITLE
Make service-account secrets in kubectl form.

### PR DESCRIPTION
**Pull request by erictune**
Changes approach taken in #5470
Instead of creating a kubernetes_auth file,
which we are tring to get away from, creates
a kubeconfig file, which is the new hotness.

Instead of creating the kubeconfig file in the
kube-addon script on the master, it creates
it at the time of salt-overlay generation.
More information is available at this time.

In particular, the master certs and master address
are handy at this point, so those are included in
the kubeconfig file.

The kube-addons script is simplified because the secret objects
are now just plain olf yaml files, which it knows how to
create, just like it creates pods and services.

Kubectl is used to generate the kubeconfig file.
This ensures correct format and is more self-documenting,
and matches how the admin credentials are done.

TODO(erictune): do this for kubelet.

### CODE REVIEW

**Comment by erictune** [comment]
@zmerlynn reviewed the last PR like this.  If he is available, he is a good reviewer for this.
@cjcullen is starting in this area and could review, and @jlowdermilk has some context too.

**Comment by zmerlynn** [comment]
Sure. I need to grab a beer anyways.

File: cluster/common.sh [hunk 1]
```diff
@@ -58,6 +58,52 @@ function create-kubeconfig() {
...
+  local token=$3
+  local context="pod"  # Any string would do here because a service account kubeconfig
+                       # is generated per-cluster and stored as a secret in that cluster.
+
+  # kubeconfig determines the file we write to, but it may not exist yet
+  if [[ ! -e "${kubeconfig}" ]]; then
+    mkdir -p $(dirname "${kubeconfig}")
+    touch "${kubeconfig}"
+  fi
+  local context="local"
```
Comment (zmerlynn) [inline cluster/common.sh]: Huh? What's going on here between this and line 79?

File: cluster/gce/configure-vm.sh [hunk 2]
```diff
@@ -268,15 +268,36 @@ function create-salt-auth() {
...
+    # The kubeconfigs go into the salt overlay so they can be used
+    # by the kube-addons component, which loads them as secrets.
     #
     # NB: If this list ever changes, this script actually has to
-    # change to detect the existence of this file, kill any deleted
+    # change to detect the known_tokens file, existence of this file, kill any deleted
     # old tokens and add any new tokens (to handle the upgrade case).
     local -r service_accounts=("system:scheduler" "system:controller_manager" "system:logging" "system:monitoring" "system:dns")
     for account in "${service_accounts[@]}"; do
+      local -r munged_username=$(tr -s ':_' '--' <<< "${account}")
```
Comment (zmerlynn) [inline cluster/gce/configure-vm.sh]: Comment what this is translating? (i.e. "system:controller_manager -> system-controller-manager")

File: cluster/gce/configure-vm.sh [hunk 3]
```diff
@@ -268,15 +268,36 @@ function create-salt-auth() {
...
+      # Generate random token
       token=$(dd if=/dev/urandom bs=128 count=1 2>/dev/null | base64 | tr -d "=+/" | dd bs=32 count=1 2>/dev/null)
+      # Put in master's list of known tokens.
       echo "${token},${account},${account}" >> "${KNOWN_TOKENS_FILE}"
+      # Generate a kubeconfig for the service account that will use this token.
+      local -r kubeconfig="/srv/salt-overlay/salt/kube-addons/${munged_username}.kubeconfig"
+      create-service-account-kubeconfig "${kubeconfig}" "${account}" "${token}"
+      # Generate a "secret" kind yaml file that embeds the kubeconfig file.
+      local -r kubeconfig_base64=$(cat "${kubeconfig}" | base64 -w0)
```
Comment (zmerlynn) [inline cluster/gce/configure-vm.sh]: This made me laugh out loud. But then I realized it's a reasonable way to roundtrip the YAML, so I guess you win. :+1:

Comment (zmerlynn) [inline cluster/gce/configure-vm.sh]: (I realize it's as-existing, it's just funnier somehow here.)

**Comment by zmerlynn** [comment]
LGTM except some nits.

File: cluster/common.sh [hunk 4]
```diff
@@ -58,6 +58,52 @@ function create-kubeconfig() {
...
+#
+#   KUBE_CERT
+#   KUBE_KEY
+#   CA_CERT
+function create-service-account-kubeconfig() {
+  local kubectl="${KUBE_ROOT}/cluster/kubectl.sh"
+  local kubeconfig=$1
+  local kube_user=$2
+  local token=$3
+  local context="pod"  # Any string would do here because a service account kubeconfig
```
Comment (jlowdermilk) [inline cluster/common.sh]: Even if any string will do, I'd recommend something like `service-account-context`.

**Comment by erictune** [comment]
This doesn't work at all.
cluster/common.sh is not available from within cluster/gce/configure-vm.sh.  The latter runs on the VM.
Also, kubectl is not available on the VM, at least at the point where this runs, which is before salt starts.
Also, the certs are not available at this point.

So, configure-vm is the wrong place to be setting this stuff up.  I am going to have to go back to the kube-addons approach.

**Comment by zmerlynn** [comment]
@erictune: Sorry for not noticing. I blame Friday afternoon.

I have a silly proposal before you go the other route. I feel like the `create-salt-auth` call is way early in `configure-vm.sh` - I put it in that order for legacy reasons when I factored all the things into functions, but honestly, there's no reason the auth files need to get created that early. I *think*, but don't quote me, that you could actually move `create-salt-auth` right before `run-salt`.

**Comment by zmerlynn** [comment]
(In fact, I suspect the logical thing is to move `ensure-kube-token` / `create-salt-pillar` / `create-salt-auth` all down to just before `run-salt`.)

**Comment by erictune** [comment]
Doing it before run-salt is still too early to use kubectl and too early to read the certs.

**Pull request closed without merging**